src/nvim/lua/treesitter.c

```diff
@@ -362,6 +375,9 @@ static int parser_parse(lua_State *L)
   buf_T *buf;
   TSInput input;
 
+  // Use redrawtime to determine the timeout
+  ts_parser_set_timeout_micros(*p, (uint64_t)p_rdt * 1000);
```

Comment:
Shouldn't the time be tracked on the Lua side. Redraw consists of multiple parsing and questing attempts. When the first parsing attempt finishes the time needed should be subtracted from the deadline of the deadline of the remaining tasks so that the total time budget doesn't get overly exceeded

Reply:
to add, subsequent parses will be quicker than the initial, but if the initial always timeouts, then it'll never finish a full parse.